In [1]:
!wget https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.tr.300.vec.gz
!gunzip cc.tr.300.vec.gz
!pip install gensim==4.3.3 numpy==1.26.4 scipy==1.13.1
from gensim.models import KeyedVectors
print("Model yükleniyor...")
model=KeyedVectors.load_word2vec_format("cc.tr.300.vec",
limit=500000)
print("Hazır!")

--2025-10-01 11:28:59--  https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.tr.300.vec.gz
Resolving dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)... 54.240.184.75, 54.240.184.45, 54.240.184.92, ...
Connecting to dl.fbaipublicfiles.com (dl.fbaipublicfiles.com)|54.240.184.75|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1261500728 (1.2G) [binary/octet-stream]
Saving to: ‘cc.tr.300.vec.gz’

cc.tr.300.vec.gz    100%[===================>]   1.17G   264MB/s    in 4.8s    

2025-10-01 11:29:04 (253 MB/s) - ‘cc.tr.300.vec.gz’ saved [1261500728/1261500728]

Model yükleniyor...
Hazır!


In [2]:
# Türkiye : Ankara :: İspanya : ?
print("Türkiye : Ankara :: İspanya : ?")
print(model.most_similar(positive=["ankara", "ispanya"], negative=["türkiye"], topn=5))
print()

# Fransa : Paris :: İtalya : ?
print("Fransa : Paris :: İtalya : ?")
print(model.most_similar(positive=["paris", "italya"], negative=["fransa"], topn=5))
print()

# Almanya : Berlin :: Japonya : ?
print("Almanya : Berlin :: Japonya : ?")
print(model.most_similar(positive=["berlin", "japonya"], negative=["almanya"], topn=5))
print()


Türkiye : Ankara :: İspanya : ?
[('italya', 0.5104808807373047), ('bursa', 0.5073603987693787), ('fransa', 0.5026073455810547), ('barcelona', 0.4961874485015869), ('Sevilla', 0.4955797791481018)]

Fransa : Paris :: İtalya : ?
[('venedik', 0.5814524292945862), ('milano', 0.570105254650116), ('pariste', 0.5643655061721802), ('Floransa', 0.5191888213157654), ('italyada', 0.5125287175178528)]

Almanya : Berlin :: Japonya : ?
[('tokyo', 0.559603214263916), ('singapur', 0.5243062973022461), ('japon', 0.5047327280044556), ('newyork', 0.4996938407421112), ('londra', 0.4907296597957611)]



#### Soru 1 Ülke - Başkent

| Analoji                         | Beklenen | Model Sonuçları (ilk 1–2)                                | Yorum                                                                                                                                |
| ------------------------------- | -------- | -------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
| Türkiye : Ankara :: İspanya : ? | Madrid   | 1. İtalya (0.51), 2. Bursa (0.50), 3. Fransa (0.50)…     | Model Madrid’i bulamadı. Corpus’ta İspanya genelde başka ülkelerle birlikte anıldığı için başkent ilişkisini kaçırmış.               |
| Fransa : Paris :: İtalya : ?    | Roma     | 1. Venedik (0.58), 2. Milano (0.57), 3. Paris’te (0.56)… | Roma yerine ünlü İtalya şehirlerini (Venedik, Milano, Floransa) getirdi. Yani ülke–başkent yerine “ülke–şehrin” ilişkisini öğrenmiş. |
| Almanya : Berlin :: Japonya : ? | Tokyo    | 1. Tokyo (0.55), 2. Singapur (0.52)…                     | Burada doğru sonucu (Tokyo) buldu . Model Japonya–Tokyo eşleşmesini öğrenmiş.                                                       |


In [6]:
# Kral : Kraliçe :: Erkek : ?
print("Kral : Kraliçe :: Erkek : ?")
print(model.most_similar(positive=["kraliçe","erkek"], negative=["kral"], topn=5))
print()

# Gündüz : Gece :: Yaz : ?
print("Gündüz : Gece :: Yaz : ?")
print(model.most_similar(positive=["gece","yaz"], negative=["gündüz"], topn=5))
print()

# Öğretmen : Okul :: Doktor : ?
print("Öğretmen : Okul :: Doktor : ?")
print(model.most_similar(positive=["okul","doktor"], negative=["öğretmen"], topn=5))
print()


# Kendi analogimiz: Ay : Gece :: Güneş : ?
print("Ay : Gece :: Güneş : ?")
result_custom = model.most_similar(positive=["gece","güneş"], negative=["ay"], topn=5)
print(result_custom)


Kral : Kraliçe :: Erkek : ?
[('kadın', 0.6786535978317261), ('kız', 0.6334383487701416), ('kadın-', 0.5685897469520569), ('erkek-kadın', 0.5654035806655884), ('bayan', 0.5578188300132751)]

Gündüz : Gece :: Yaz : ?
[('Yaz', 0.6220954060554504), ('kış', 0.6100205183029175), ('sonbahar', 0.5917434692382812), ('bahar', 0.5499704480171204), ('ilkbahar', 0.5494310259819031)]

Öğretmen : Okul :: Doktor : ?
[('hastane', 0.728849470615387), ('hastahane', 0.6638420224189758), ('hastanenin', 0.5605310201644897), ('hastaneler', 0.540983259677887), ('hastahaneler', 0.5363527536392212)]

Ay : Gece :: Güneş : ?
[('gündüz', 0.6085719466209412), ('güneşin', 0.5537223219871521), ('Gece', 0.5426763296127319), ('güneşiyle', 0.5351989269256592), ('güneşinin', 0.5309818387031555)]


### Soru 2 – Kavramsal İlişkiler

**Kral : Kraliçe :: Erkek : ?**

Model doğru şekilde “kadın” sonucunu ilk sırada bulmuş.

Bu, modelin cinsiyet karşılıklarını ve toplumsal cinsiyet ilişkilerini iyi öğrendiğini gösteriyor.

Ayrıca “kız”, “bayan” gibi varyasyonların da çıkması, modelin eşanlamlıları da yakalayabildiğini gösteriyor.

**Gündüz : Gece :: Yaz : ?**

Model mevsim ve zaman ilişkilerini kavrayabiliyor.

İlk sırada “yaz” çıkmış olsa da, ikinci sırada doğru cevap “kış” mevcut.

Bu durum, modelin bazı bağlamlarda frekansa bağlı olarak farklı sonuçlar üretebileceğini gösteriyor.

**Öğretmen : Okul :: Doktor : ?**

Model bağlamsal ilişkileri çok başarılı şekilde yakalamış; “hastane” ve türevleri en üst sıralarda.

Bu, meslek ve mekan ilişkilerini öğrenebildiğini gösteriyor.

Farklı kelime varyasyonlarının çıkması, modelin geniş bir kelime kullanım bilgisini kapsadığını gösteriyor.

**Ay : Gece :: Güneş : ?**

Model doğru şekilde “gündüz” sonucunu ilk sırada bulmuş

Bu, modelin doğa olayları ve zaman kavramları arasındaki ilişkiyi kavrayabildiğini gösteriyor.

Ayrıca “güneşin”, “güneşiyle” gibi varyasyonların da çıkması, modelin kelime türevlerini ve bağlam ilişkilerini yakalayabildiğini gösteriyor.


In [5]:
# Hemşire : Kadın :: Doktor : ?
print("Hemşire : Kadın :: Doktor : ?")
print(model.most_similar(positive=["kadın","doktor"], negative=["hemşire"], topn=5))
print()


Hemşire : Kadın :: Doktor : ?
[('kadının', 0.6143438816070557), ('erkek', 0.6107356548309326), ('kadınlar', 0.5653315782546997), ('jinekolog', 0.5632423758506775), ('kadını', 0.5555832982063293)]



### Soru 3 –  Bias Gözlem

Model, hemşire–kadın ve doktor–erkek eşleşmelerini veri üzerinden öğrenmiş görünüyor.

Çıkan ilk sonuçların “kadının” ve “kadınlar” olması, modelin toplumsal cinsiyet önyargısını yansıttığını gösteriyor.

Bu, word embedding modellerinin eğitim verisindeki önyargıları aynen öğrenip sonuçlara taşıyabileceğini ortaya koyuyor.

In [7]:
# 1️⃣ Elma : Meyve :: Havuç : ?
print("Elma : Meyve :: Havuç : ?")
result1 = model.most_similar(positive=["meyve","havuç"], negative=["elma"], topn=5)
print(result1)
print()

# 2️⃣ Kitap : Kütüphane :: Para : ?
print("Kitap : Kütüphane :: Para : ?")
result2 = model.most_similar(positive=["kütüphane","para"], negative=["kitap"], topn=5)
print(result2)
print()

# 3️⃣ Beşiktaş : İstanbul :: Barcelona : ?
print("Beşiktaş : İstanbul :: Barcelona : ?")
result3 = model.most_similar(positive=["istanbul","barcelona"], negative=["beşiktaş"], topn=5)
print(result3)
print()


Elma : Meyve :: Havuç : ?
[('sebze', 0.7243412137031555), ('sebzeler', 0.6214406490325928), ('sebzelere', 0.6159278750419617), ('meyve-sebze', 0.5970016121864319), ('sebze-meyve', 0.5965595841407776)]

Kitap : Kütüphane :: Para : ?
[('paralar', 0.48872601985931396), ('parayı', 0.47033238410949707), ('paraları', 0.46755722165107727), ('birimi', 0.46665671467781067), ('döviz', 0.4539383351802826)]

Beşiktaş : İstanbul :: Barcelona : ?
[('ankara', 0.536542534828186), ('londra', 0.5338242053985596), ('ispanya', 0.47953665256500244), ('tokyo', 0.47438669204711914), ('italya', 0.4721299111843109)]



## Soru 4 – Kompleks İlişkiler

#### 1️ Elma : Meyve :: Havuç : ?
**Model Sonuçları:** Sebze (0.724), Sebzeler (0.621), Sebzelere (0.616), Meyve-sebze (0.597), Sebze-meyve (0.597)  
**Yorum:**  Model doğru sonucu bulmuş: Havuç → Sebze. Türevler de çıkmış, kategori ilişkileri iyi öğrenilmiş.

#### 2️ Kitap : Kütüphane :: Para : ?
**Model Sonuçları:** Paralar (0.489), Parayı (0.470), Paraları (0.468), Birimi (0.467), Döviz (0.454)  
**Yorum:**  Beklenen “Banka” yerine ekonomik bağlam kelimeleri çıkmış. Model bağlamı tam anlayamamış.

#### 3️ Beşiktaş : İstanbul :: Barcelona : ?
**Model Sonuçları:** Ankara (0.537), Londra (0.534), İspanya (0.480), Tokyo (0.474), İtalya (0.472)  
**Yorum:**  Beklenen “İspanya” üçüncü sırada çıkmış. Model, Barcelona’yı daha çok futbol veya şehir bağlamında öğrenmiş.

#####  Genel Tartışma
- Model bazı ilişkileri doğru, bazılarını bağlam farkı nedeniyle yanlış veya gecikmeli çıkarmış.  
- Doğru çıkmayan örnekler genellikle:  
  - Kelimenin corpus’ta az veya farklı bağlamlarda geçmesi  
  - Çok anlamlı kelimelerin birden fazla bağlamda öğrenilmesi  
  - Modelin istatistiksel co-occurrence bazlı çalışması


In [8]:
# Kalem : Yazmak :: Fırça : ?
result_custom2 = model.most_similar(positive=["yazmak","fırça"], negative=["kalem"], topn=5)
print(result_custom2)


[('yayınlamak', 0.526287853717804), ('yazmamak', 0.5225868821144104), ('yorumlamak', 0.5053537487983704), ('okumak', 0.4803018867969513), ('Yazmak', 0.4748064875602722)]


### Soru 5 – Yaratıcı Analogi

#### Kalem : Yazmak :: Fırça : ?
**Model Sonuçları:** Yayınlamak (0.526), Yazmamak (0.523), Yorumlamak (0.505), Okumak (0.480), Yazmak (0.475)  
**Beklenen Cevap:** Boyamak  
**Tartışma ve Yorum:**  
Model beklenen “boyamak” sonucunu verememiş. Bunun sebebi, embedding modelinin “fırça” kelimesini corpus’ta çoğunlukla yazı, sanat veya sanat yorumlama bağlamında görmesi olabilir.  
Çıkan sonuçlar (“yayınlamak”, “yorumlamak”, “okumak”) daha çok metin ve yayıncılık bağlamıyla ilişkili. Bu, modelin kelimenin kullanım bağlamına bağlı olduğunu gösteriyor.  
“Yazmak” ve “yazmamak” gibi kelimeler, modelin temel eşleştirme mantığını kavradığını ama araç–işlev ilişkisini yeterince öğrenemediğini gösteriyor.  
Bu durum, embedding modellerinin somut araç–işlev ilişkilerinde sınırlı kalabileceğini ve sonuçların her zaman mantıklı çıkmayabileceğini ortaya koyuyor.
